pkg/pillar/cmd/zedkube/kubenodeop.go

```diff
@@ -0,0 +1,267 @@
+
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{"node-uuid": ctx.nodeuuid}}
+	options := metav1.ListOptions{LabelSelector: metav1.FormatLabelSelector(&labelSelector)}
+	nodes, err := clientset.CoreV1().Nodes().List(context.Background(), options)
+	if err != nil {
+		return fmt.Errorf("drainNode: can't get nodes %v, on uuid %s", err, ctx.nodeuuid)
+	}
+	if len(nodes.Items) == 0 {
+		return fmt.Errorf("drainNode: can't find node")
+	}
+	node := nodes.Items[0]
+	nodeName := node.Name
+
+	// https://longhorn.io/docs/archives/1.4.0/volumes-and-nodes/maintenance/#updating-the-node-os-or-container-runtime
+	// https://github.com/longhorn/longhorn/discussions/8593
+	lhPodSelectors := []string{
+		"app!=csi-attacher",
+		"app!=csi-provisioner",
+		"app!=longhorn-admission-webhook",
+		"app!=longhorn-conversion-webhook",
+		"app!=longhorn-driver-deployer",
+	}
+	podSelectorStr := strings.Join(lhPodSelectors, ",")
+
+	drainHelper := &drain.Helper{
+		Client:              clientset,
+		Force:               true,
+		GracePeriodSeconds:  -1,
+		IgnoreAllDaemonSets: true,
+		Out:                 os.Stdout,
+		ErrOut:              os.Stderr,
+		DeleteEmptyDirData:  true,
+		Timeout:             time.Hour * time.Duration(ctx.drainTimeoutHours),
+		PodSelector:         podSelectorStr,
+	}
+	err = drain.RunNodeDrain(drainHelper, nodeName)
+	if err != nil {
+		drainErr := fmt.Errorf("drainNode RunNodeDrain Failure: %v", err)
+		log.Error(drainErr)
+		return drainErr
+	}
+	log.Noticef("drainNode: node %s drained", nodeName)
+	return nil
+}
+
+// try to drain and delete the node before we remove the cluster config and
+// transition into single-node mode. Otherwise, if the node is later added to
+// the cluster again, it will not be allowed due to duplicate node names.
+func drainAndDeleteNode(ctx *zedkubeContext) {
+	config, err := kubeapi.GetKubeConfig()
+	if err != nil {
+		log.Errorf("drainAndDeleteNode: can't get kubeconfig %v", err)
+		return
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Errorf("drainAndDeleteNode: can't get clientset %v", err)
+		return
+	}
+
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{"node-uuid": ctx.nodeuuid}}
+	options := metav1.ListOptions{LabelSelector: metav1.FormatLabelSelector(&labelSelector)}
+	nodes, err := clientset.CoreV1().Nodes().List(context.Background(), options)
+	if err != nil {
+		log.Errorf("drainAndDeleteNode: can't get nodes %v, on uuid %s", err, ctx.nodeuuid)
+		return
+	}
+	if len(nodes.Items) == 0 {
+		log.Errorf("drainAndDeleteNode: can't find node")
+		return
+	}
+	node := nodes.Items[0]
+	nodeName := node.Name
+
+	// cordon the node first
+	node.Spec.Unschedulable = true
+	_, err = clientset.CoreV1().Nodes().Update(context.Background(), &node, metav1.UpdateOptions{})
+	if err != nil {
+		log.Errorf("drainAndDeleteNode: cordon node %s failed: %v, continue the delete", nodeName, err)
+	}
+
+	if err = drainNode(ctx); err != nil {
```

Comment:
the drainNode() seems only handle the longhorn related pods, not 'eve-kube-app' related pods, we should delete those pods before delete the node.

Reply:
@naiming-zededa Added deletePods() in kubenodeop.go called in drainAndDeleteNode() just after cordoning the node.